coderd/notifications/reports/generator.go

```diff
@@ -0,0 +1,296 @@
+package reports
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"io"
+	"slices"
+	"sort"
+	"time"
+
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+
+	"github.com/google/uuid"
+
+	"github.com/coder/quartz"
+
+	"github.com/coder/coder/v2/coderd/database"
+	"github.com/coder/coder/v2/coderd/database/dbauthz"
+	"github.com/coder/coder/v2/coderd/database/dbtime"
+	"github.com/coder/coder/v2/coderd/notifications"
+	"github.com/coder/coder/v2/codersdk"
+)
+
+const (
+	delay = 15 * time.Minute
+)
+
+func NewReportGenerator(ctx context.Context, logger slog.Logger, db database.Store, enqueuer notifications.Enqueuer, clk quartz.Clock) io.Closer {
```

Comment:
I don't see any tests which exercise this, which makes me wonder how you've validated the locking?

Reply:
You are correct, I haven't. Do you have a specific idea of what unit test would fulfill the gap? I could run two instances of Report Generator, but then what would we like to validate and how?

Reply:
Running multiple instances sounds good like a good approach.
You should aim to achieve a moderately high % of coverage since this is a crucial area. Achieving the coverage will require provoking the various corner-cases, like when a lock is already held when trying to run a report.

Reply:
I see what you mean, but I have no clue how we can test locking without some _blackbox/timeout_ tests or artificially mocked database calls. I'm happy to pair-program on this with you.

IMHO the most crucial part of this implementation is making sure that consecutive runs (immediate runs) won't send reports. What is not tested using the current approach is exclusive locking, and this will be hard.

Reply:
> What is not tested using the current approach is exclusive locking, and this will be hard.

What makes this particularly hard (asking for real, not rhetorically)? To me it seems like we'd need to invoke `NewReportGenerator` in two separate goroutines and just ensure that we only get what we're expecting, which you've implemented in other tests.

I think it's quite risky to have functionality which should be guarding against pathological behaviour which is not covered by a test.